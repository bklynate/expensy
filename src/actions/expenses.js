/* eslint-disable arrow-body-style */
// import uuid from 'uuid';
import { db } from './../firebase/firebase';

// Action List - Create action generators for these:

// component calls an action generator, which in turn returns an object
// the component then dispatchs an object to the redux store and it changes

// async redux
// component calls action generator
// action generator returns function
// component dispatches function
// function runs and has the ability to cause other actions

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => dispatch => {
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = expenseData;
  const expense = { description, note, amount, createdAt };
  return db.ref('expenses').push(expense).then(ref => {
    dispatch(
      addExpense({
        id: ref.key,
        ...expense
      })
    );
  });
};
// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  expense: { id }
});

export const startRemoveExpense = ({ id }) => {
  return (dispatch) => {
    return db.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }))
    })
  }
}
// EDIT_EXPENSE
export const editExpense = ({ id, updates }) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = ({ id, updates }) => {
  return (dispatch) => {
    return db.ref(`expenses/${id}`).update({
      ...updates
    }).then(() => {
      dispatch(editExpense({
        id,
        updates
      }))
    })
  }
}

// SET_EXPENSE
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch) => {
    return db.ref('expenses').once('value')
      .then(data => {
        const expenses = [];
        data.forEach(item => {
          expenses.push({
            id: item.key,
            ...item.val()
          })
        })
        dispatch(setExpenses(expenses));
      })
  }
};
