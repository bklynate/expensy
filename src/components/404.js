import React from 'react';
import { Link } from 'react-router-dom';

const ErrorPage = () => (
  <div>
    <h1>
      404 Error - <Link to="/">Go Home</Link>
    </h1>
  </div>
);

export default ErrorPage;
