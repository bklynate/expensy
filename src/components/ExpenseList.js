import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import ExpensesSummary from './ExpenseSummary';
import expenseSelector from './../selectors/expenses';
import expenseTotal from './../selectors/expenseTotal';

export const ExpenseList = props => (
  <div>
    <h1>Expense List</h1>
    <ExpensesSummary />
    {props.expenses.length ? (
      props.expenses.map(({ id, description, amount, createdAt }) => (
        <ExpenseListItem
          key={id}
          id={id}
          description={description}
          amount={amount}
          createdAt={createdAt}
        />
      ))
    ) : (
      <p>There are currently {props.expenses.length} expenses to track.</p>
    )}
  </div>
);

const mapStateToProps = state => ({
  expenses: expenseSelector(state.expenses, state.filters),
  total: () => expenseTotal(state.expenses)
});

export default connect(mapStateToProps)(ExpenseList);
