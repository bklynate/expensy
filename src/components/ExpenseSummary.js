import React from 'react';
import { connect } from 'react-redux';
import expenseSelector from './../../src/selectors/expenses';
import expenseTotal from './../../src/selectors/expenseTotal';

export const ExpenseSummary = props => (
  <div>
    <h2>
      {props.expenses.length
          ? `Your current expenses equal: ${props.total(props.expenses)}`
          : `You have ${props.total()} in expenses`}
    </h2>
  </div>
);

const mapStateToProps = state => ({
  expenses: expenseSelector(state.expenses, state.filters),
  total: () => expenseTotal(expenseSelector(state.expenses, state.filters))
});
export default connect(mapStateToProps)(ExpenseSummary);
