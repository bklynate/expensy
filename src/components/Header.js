import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <div>
    <h1>Xpensy</h1>
    <NavLink exact activeClassName="is-active" to="/">
      DashBoard
    </NavLink>
    <NavLink activeClassName="is-active" to="/create">
      Create Expense
    </NavLink>
    <NavLink activeClassName="is-active" to="/help">
      Help
    </NavLink>
  </div>
);

export default Header;
