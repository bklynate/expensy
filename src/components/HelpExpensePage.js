import React from 'react';

const HelpExpensePage = () => (
  <div>
    <h1>How does expensy work?</h1>
  </div>
);

export default HelpExpensePage;
