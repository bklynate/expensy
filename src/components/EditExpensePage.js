import React, { Component } from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from './../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends Component {
  onSubmit = expense => {
    this.props.editExpense(expense.id, expense);
    this.props.history.push('/');
  };

  onRemove = () => {
    const { id } = this.props.expense
    this.props.removeExpense({ id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  editExpense: expense => dispatch(editExpense(expense.id, expense)),
  removeExpense: ({ id }) => dispatch(removeExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
