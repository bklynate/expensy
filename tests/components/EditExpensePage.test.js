import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from './../../src/components/EditExpensePage';
import expenses from './../fixtures/expenses';

let editExpense;
let removeExpense;
let history;
let wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      history={history}
      removeExpense={removeExpense}
      expense={expenses[1]}
    />
  );
});

describe('EditExpensePage Component', () => {
  it('should EditExpensePage component properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle onSubmit properly', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
  });

  it('should handle onRemove properly', () => {
    const { id } = expenses[1];
    wrapper.find('button').simulate('click');
    expect(removeExpense).toHaveBeenLastCalledWith({ id });
    expect(history.push).toHaveBeenLastCalledWith('/');
  });
});
