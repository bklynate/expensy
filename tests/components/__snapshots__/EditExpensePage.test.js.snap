// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage Component should EditExpensePage component properly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 34589,
        "createdAt": 0,
        "description": "November gas payment",
        "id": "2",
        "note": "This is my last gas payment for November",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
